highfidelity/hifi
# Material Entities

Adds a new entity type, Material, which can be used to modify the existing materials on model entities, shape entities (albedo only), model overlays, and avatars.

Material entities have the following properties:
- `materialURL` (string):  URL to a material description.  Currently, only JSON format is supported.  Set to "userData" to use the user data to live edit a material.  The following fields are supported in the JSON:
  - "materialVersion": version of this network material (defaults to 1)
  - "materials": an object or an array of objects, each with the following properties:
    - strings:
        - name (not yet used), model (not yet used, defaults to "hifi_pbr")
    - floats:
        - opacity, roughness, metallic, scattering
    - bool:
        - unlit
    - colors (arrays of 3 floats 0-1.  Optional fourth value in array can be a boolean isSRGB):
        - emissive, albedo
    - urls to textures:
        - emissiveMap, albedoMap (set opacityMap = albedoMap for transparency), metallicMap or specularMap, roughnessMap or glossMap, normalMap or bumpMap, occlusionMap, lightmapMap (not working, currently), scatteringMap (only works if also normal mapped)
- `materialMappingMode` (string): "uv" or "projected", but currently only "uv" is supported.
- `priority` (uint):  Priority for this material when applying it to its parent.  Only the highest priority material will be used.  Materials with the same priority are (essentially) randomly sorted.  Base materials that come with models always have priority 0.
- `parentMaterialName` (string or uint):  An identifier for choosing a submesh or submeshes within a parent.  If in the format "mat::string", all submeshes with material name "string" will be replaced.  Otherwise, parentMaterialName will be parsed as an unsigned int (strings not starting with "mat::" will parse to 0), representing the mesh index to modify.  You only need to add "mat::" if setting parentMaterialName from a script;  the Create app does it for you.
- `materialMappingPos` (vec2):  Offset position in UV-space of top left of material, (0, 0) to (1, 1)
- `materialMappingScale` (vec2):  How much to scale this material within its parent's UV-space
- `materialMappingRot` (float):  How much to rotate this material within its parent's UV-space (degrees)

Material entities render as (non-scale-able) spheres if they don't have their parent set.

To apply a material to an entity or overlay, set the material entity's `parentID` to the EntityID or OverlayID of the object.

To apply a material to an avatar, create an Material Avatar Entity (`Entity.addEntities(props, true)`);  In this case, the parent is set with `owningAvatarID` instead of `parentID`.

You can append "?string" to the `materialURL` set the active material from the JSON used by the material entity.  The material with name "string" in "materials" will be selected, if it exists.

Interface script for applying an animated material to yourself: https://gist.githubusercontent.com/SamGondelman/bfa541c3ccfc81955dcfeb78a7553cb1/raw/57154b32bf36ee359b8e48d8846305fd68d602fd/MaterialAnimation.js
Entity script for applying material on grab (attach to a material, grab it, y position controls submesh): https://gist.githubusercontent.com/SamGondelman/5d352cd47f50082f727221ba22970ef5/raw/259d5b21a969fec604bfbd581b430ebc48f9147c/ApplyMaterial.js

# Test plan
- Run all of the material entity create/apply tests from hifi_tests: https://github.com/SamGondelman/hifi_tests/tree/4e777a9fe6296b87a0be113e366df0a19c21d71c/tests/content/entity/material (corresponding to https://github.com/highfidelity/hifi_tests/pull/69)
- Verify that you can use the Create app to make a material entity.  You can use this example materialURL, or create your own: https://hifi-content.s3.amazonaws.com/samuel/TestMaterial.json
- Verify that you can edit the properties of that material entity and that the changes are reflected on the sphere as expected.
- In the entity list, you should be able to see the material URL next to the material.
- Create two materials with materialURL "userData".  Set their parent to be the same model entity.  Set their albedos in their userDatas (e.g. `{ "materials": {"albedo": [1, 0, 0] } }`) to two different colors.  Play around with their priorities.  The highest priority one should be displayed.  If they have the same priority, it is basically random which one will take effect.
- Change the model's URL to a different model.  Once the new model loads, the materials should still be applied to it.
- Delete the two material entities.  The model's color should return to normal.
- Create a Cube entity and edit its color using Create.  It should work as expected.
- Create a material entity with url https://hifi-content.s3.amazonaws.com/samuel/TestMaterial2.json.  It should be red.  Add "?blue" to the end of the url.  It should be blue.  Replace "blue" with "green".  It should be green.
